docs/design-proposals/2023-11 TraceQL Metrics.md

```diff
@@ -0,0 +1,112 @@
+---
+Authors: Martin Disibio (@mdisibio)
+Created: 2023 November
+Last updated: 2023-11-13
+---
+
+# TraceQL Metrics
+
+## Summary
+
+This design document describes TraceQL language additions for extracting time-series metrics like error rates and latencies from traces.   
+
+This document is **not** meant to be a complete language specification. Rather it is meant to invite the community to comment on the direction and capabilities.
+
+### Table of Contents
+
+- [Intro](#intro)
+- [First Stage](#first-stage)
+  - [Interval](#interval)
+  - [Grouping](#grouping)
+  - [Filtering](#filtering)
+- [Additional Stages](#additional-stages)
+- [Arithmetic](#arithmetic)
+- [Notes](#notes)
+  - [Step Interval](#step-interval)
+
+## Intro
+
+Let's look at an example straight away.  This query plots the error rate by service over time:  
+`{status = error } | rate() by (resource.service.name)`
```

Comment:
I believe adopting the logQL/promQL syntax for aggregations in traceQL would make it more familiar to the users and straightforward to adopt.

Current proposal:
`{status = error } | rate() by (resource.service.name)`

Suggested logQL/promQL Style:
`sum(rate({status = error })) by (resource.service.name)`

Reply:
Yep this is a good called and we invested a lot of thought here but I didn't cover it in the proposal.  Ultimately decided that Tempo can't execute these basic queries and it would require breaking changes.  If if breaking changes are required, then even better to use the pipeline structure rather than a close-but-different promql syntax.

Here is some explanation:

`sum(rate({status = error }))`

Inside this query is:

`rate({status =error})` which returns the raw series in promql/logql. These series have implied labels which are a side effect of your application instrumentation and your data pipeline.  For instance your app might include the label `userID` and your pipeline might include the label `region`.  The labels are a fundamental design aspect of the data and your infrastructure.   Tempo has no implied labels and doesn't require decisions about them or limits to cardinality.  So it's not clear what to return from the query `rate({status=error})`.  

If it returns the total rate, there is no need for `sum()`. 

If it returns "raw series" it would have to pull something from the span/resource/trace (pulling "everything" isn't feasible or useful).  The `by()` clause contains this information but it must now be attached to `rate` instead of `sum`.  

This is how we arrived at `... | rate() by ()`.

Reply:
Thanks for the detailed explanation. I understand the team concerns about the breaking changes and the unique aspects of Tempo.

To clarify my suggestion: My expectation was that `rate` would return all intrinsics, along with span/resource attributes, similar to Loki/logQL's behavior. This approach might not be useful in isolation, but it becomes powerful when combined with `sum`.

For instance, in the **intro-to-mlt** project, using the feature `log_received_spans` this raw data is overwhelming in Loki but when we apply the `sum` function to group this data, it transforms into actionable insights.

<img width="1913" alt="image" src="https://github.com/grafana/tempo/assets/2423149/3701704f-cf1a-4e9a-a1ee-310f940f6a97">

![image](https://github.com/grafana/tempo/assets/2423149/d02f9a11-c6bb-4766-a4d9-e544572e73d6)

I understand that implementing this would require significant changes. However, I believe aligning traceQL more closely with familiar syntaxes could lower the learning curve.

Reply:
@faustodavid To be clear the proposed traceql `rate()` func is already the summed total.  

Traceql query for the second screenshot would be:

`{status=error && name="pg.query:INSERT postgres"} | rate() by (span.db.statement)`